runtime/desired.go

```diff
@@ -0,0 +1,198 @@
+package runtime
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	xkube "github.com/crossplane-contrib/provider-kubernetes/apis/object/v1alpha1"
+	xfnv1alpha1 "github.com/crossplane/crossplane/apis/apiextensions/fn/io/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type DesiredResources struct {
+	resources []Resource
+	composite xfnv1alpha1.DesiredComposite
+}
+
+// GetFromKubeObject gets the k8s resource o from a provider kubernetes object kon (Kube Object Name)
+// from the desired array of the FunctionIO.
+func (d *DesiredResources) GetFromKubeObject(ctx context.Context, o client.Object, kon string) error {
+	ko, err := getKubeObjectFrom(ctx, &d.resources, kon)
+	if err != nil {
+		return fmt.Errorf("cannot get resource from kube object: %v", err)
+	}
+	return fromKubeObject(ctx, ko, o)
+}
+
+// ResourceExists check weather a relevant resource exists in this slice.
+// A relevant resource is any resource that is not a Kubernetes Object resource.
+// The function also checks resources inside Kubernetes Objects in case unmarshalling
+// does not fail.
+func (d *DesiredResources) ResourceExists(ctx context.Context, name string) bool {
```

Comment:
I'm not sure if this is what we want and IMO is a bit too much magic.

One edge case I can see if there is a kube object that contains a resource called `foo`, but I actually want to check if a managed resource `foo` exits, this will return `true`. I.e there can be clashes with kube object names and managed resource names.

I'm not convinced we actually need this method. One can just call `Get` and see if it fails.

Reply:
I introduced this method during test refactoring and I thought it might be useful in the long run. Currently, after several refactoring of unit tests I ended up not using it anymore. Moreover you made a valid point there so I will remove it.